docs_src/design/adr/system-management/0006-Metrics-Collection.md

```diff
@@ -0,0 +1,382 @@
+m := metrics.NewMeter()
+m.Mark(1)  // add one to the current meter value
+time.Sleep(15 * time.Second)  // allow some time to go by
+m.Mark(1)  // add one to the current meter value
+time.Sleep(15 * time.Second)  // allow some time to go by
+m.Mark(1)  // add one to the current meter value
+time.Sleep(15 * time.Second)  // allow some time to go by
+m.Mark(1)  // add one to the current meter value
+time.Sleep(15 * time.Second)  // allow some time to go by
+fmt.Println(m.Count())  // prints 4
+fmt.Println(m.Rate1())  // prints 0.11075889086811593
+fmt.Println(m.Rate5())  // prints 0.1755318374350548
+fmt.Println(m.Rate15()) // prints 0.19136522498856992
+fmt.Println(m.RateMean()) //prints 0.06665062941438574
+```
+
+- Histograms: measure the statistical distribution of values (int64 values) in a collection of values.
+    - Example use: response times on APIs
+    - Operations: update and get the min, max, count, percentile, sample, sum and variance from the collection
+
+``` Go
+h := metrics.NewHistogram(metrics.NewUniformSample(4))
+h.Update(10)
+h.Update(20)
+h.Update(30)
+h.Update(40)
+fmt.Println((h.Max()))  // prints 40
+fmt.Println(h.Min())    // prints 10
+fmt.Println(h.Mean())   // prints 25
+fmt.Println(h.Count())  // prints 4
+fmt.Println(h.Percentile(0.25))  //prints 12.5
+fmt.Println(h.Variance())  //prints 125
+fmt.Println(h.Sample())  //prints &{4 {0 0} 4 [10 20 30 40]}
+```
+- Timer: measures both the rate a particular piece of code is called and the distribution of its duration
+    - Example use:  how often an app service function gets called and how long it takes get through the function
+    - Operations:  update and get min, max, count, rate1, rate5, rate15, mean, percentile, sum and variance from the collection
+
+``` Go
+t := metrics.NewTimer()
+t.Update(10)
+time.Sleep(15 * time.Second)
+t.Update(20)
+time.Sleep(15 * time.Second)
+t.Update(30)
+time.Sleep(15 * time.Second)
+t.Update(40)
+time.Sleep(15 * time.Second)
+fmt.Println((t.Max()))  // prints 40
+fmt.Println(t.Min())    // prints 10
+fmt.Println(t.Mean())   // prints 25
+fmt.Println(t.Count())  // prints 4
+fmt.Println(t.Sum())    // prints 100
+fmt.Println(t.Percentile(0.25))  //prints 12.5
+fmt.Println(t.Variance())  //prints 125
+fmt.Println(t.Rate1())  // prints 0.1116017821771607
+fmt.Println(t.Rate5())  // prints 0.1755821073441404
+fmt.Println(t.Rate15()) // prints 0.1913711954736821
+fmt.Println(t.RateMean()) //prints 0.06665773963998162
+
+```
+
+!!! Note
+    The go-metrics package does offer some variants of these like the GaugeFloat64 to hold 64 bit floats.
+
+## Consequences
+
+- Should there be a *global* configuration option to turn all metrics off/on?
+    - * EdgeX doesn't yet have global config so this will have to be by service. *
+- Given the potential that each service publishes metrics to the same message topic, 0MQ is not implementation option unless each service uses a different 0MQ pipe (0MQ topics do not allow multiple publishers).  

```

Comment:
Since we've deprecated ZMQ, shouldn't this bit be removed?

Reply:
@tonyespy,  ZMQ is still an option, just deprecated, so IMO this should stay so folks know they they can not use ZMQ with these Metrics.